### ISSUE TITLE
Add Keyboard events

### ISSUE THREAD

**Issue by harinikmsft** [microsoft/react-native-windows#2469]
Creating a separate issue to track the APIs needed for overall keyboarding and routed events support in react-native-windows. 

**Scenarios:**

- This is needed before we can solve issues tracked by #2085 , #2107 and perhaps #2197 . 
- This is also needed in order for apps to enable custom keyboarding behaviors at any layer, for example - ability to handle arrow keys, Esc key strokes from a custom TextInput component to wire up custom behaviors before the framework executes a default behavior. 

**Scope:**

- Introduce APIs like [ReactJS Keyboard events](https://reactjs.org/docs/events.html#keyboard-events) in react-native-windows. 
- Do they map to native UWP [Keyboard events](https://docs.microsoft.com/en-us/windows/uwp/design/input/keyboard-events) in implementation?
- At what layer are these events exposed?
- Spec the behavior expectations from a threading model, bubbling/tunneling nature of these routed events in react-native-windows.
- Will this be specific to Windows or can this be implemented in a platform agnostic manner and pulled to react-native?

## [Updated 5/30] 
**Proposal:**

The spec/proposal for this has work has been captured [here](https://github.com/ReactWindows/discussions-and-proposals/blob/harinikmsft-keyboardapis/proposals/0000-keyboard-apis.md) . This has passed initial review and we can go ahead with the implementation for now. There are a few outstanding broader eventing threading model related issues that need to be resolved with the RN/React community which we shall be doing in parallel. Once we feel confident about the shape of this API, the spec/proposal draft will be PR'ed into react-native master fork for broader feedback and adding support for other platforms as well.

**Comment by harinikmsft** [comment]
Proposal for this work is being tracked through the PR here - [Keyboard APIs](https://github.com/ReactWindows/discussions-and-proposals/pull/1)

**Comment by kmelmon** [comment]
Load balancing this over to CanHua

**Comment by kmelmon** [comment]
Here's some info from an email thread that might be useful for this task:

I touched on this a little in our meeting, but the behavior of the SyntheticEvent wrapper on RN (henceforth NativeSyntheticEvent) is a little strange (and I think Facebook is likely to change some of this when they do their Fire & Fabric rewrites).  I don't know the exact mechanics of how SyntheticEvent behaves in ReactJS, I didn't investigate that deeply, though I could probably loop in some ReactJS experts on our team if you're interested in that information.  To be clear, I just learned about this behavior through ad hoc experimentation and reverse engineering a while ago, so take some of this with a grain of salt.

Here's what I know:

Inside of NativeSyntheticEvent is a nativeEvent field, which is a JSON blob defined by the event callback prop in question.  The various fields inside of NativeSynetheticEvent appear to just reflect the corresponding field in the JSON blob if it exists, otherwise its just left undefined.  That is, if NativeSyntheticEvent.nativeEvent.bubbles is true, then NativeSyntheticEvent.bubbles will also be true.  This data is not inferred by the callbacks being exposed via GetExportedCustomBubblingEventTypeConstants and must be explicitly dispatched across the bridge along with whatever other event args you want to include in your nativeEvent.  I found it very curious that this wrapper is so behavior-less by default.  

There are two behaviors I did notice from the wrapper that weren't just transparent pass-throughs from the nativeEvent:
•	If you want to stop event propagation on the JS side, you can call preventDefault (or stopPropagation?) like you would in regular DOM+JS coding.
•	If you happen to include "target" on the nativeEvent, then NativeSyntheticEvent.currentTarget will be automatically updated for events that bubble on the JS side.
Regarding the eventPhase field-- like bubbles, it is undefined unless included in nativeEvent and I don't remember it changing during JS-side event propagation.  This seems quite useless to me, because the native side cannot know (or change) the JS-side event phase and I'm not sure how much the JS-side cares about the native event phase after the event has been dispatched.  Perhaps Facebook would agree this is a bug, or maybe they just don't expect people to use the fields in the NativeSyntheticEvent... I'm guessing SyntheticEvent in ReactJS is a little bit more well defined in this regard because of how much it aligns with the DOM event structure.

All that said I feel like I need to reiterate something I stated in my previous email just because I'm not sure if it came across clearly:

I added the eventPhase enum for the declarative prop to control when to mark native events as handled during native event propagation.  eventPhase in the NativeSyntheticEvent seems useless in its present state, ideally I think the RN framework code would update it with the current JS event phase as event propagation happens on the JS side and perhaps even allow the NativeSyntheticEvent.nativeEvent.eventPhase to differ from NativeSyntheticEvent.eventPhase (which would refer to the native event phase that originally dispatched the event and the current JS event phase, respectively).

Barring synchronous communication promised in the Fabric rewrite which could maybe allow event propagation to happen in both sides in lockstep (??) or some other form of synchronous native event handling, I think the eventPhase I introduced for the prop serves a fundamentally different purpose than the one in the event args in the callback.  However, I aligned with the naming conventions the web uses for its eventPhases in its callbacks to describe the native eventPhase of the prop.  I made all the fields in the prop object-array optional except Key, which means eventPhase undefined or None is valid and I'd be curious to know what you folks thing is an intuitive default behavior, there.  

I haven't had an opportunity to read your proposal yet, so I don't know if you're going in a different direction, but I hope that helps!

Alicia
________________________________________
From: Harini Kannan
Sent: Wednesday, May 22, 2019 3:14 PM
To: Alicia Drummond; Andrew Coates (REDMOND); Mike Hillberg; Kiki Saintonge
Cc: Keith Melmon; Stephen Peters
Subject: RE: onKey*Capture events 
 
+@Mike Hillberg and @Kiki Saintonge (a lot of these are concepts that will apply to Pointer* events too since they are also routed).
 
Thanks Alicia. That makes a lot of sense. I am editing my PR to reflect some of this.
 
You are right about the WPF routing phases. I considered calling the capture events as previewKeyDown/previewKeyUp to match the windows APIs more, but it does look like React/JS has a more general idea of appending any event with “Capture” to handle the event in the capturing phase rather than the bubbling phase (see the blurb in this section). So the API you have landed seems to be definitely more appropriate for RN. 
 
Question though – it appears from reading through React events and W3 events docs that eventPhase is more of a common property for all events and there are many associated properties that go along with it like Event.bubbles, Event.Cancelable etc., Is this not available as part of the ReactNative.NativeSyntheticEvents and is it the right model for us to be adding this explicitly to the IKeyboardEvent instead of at a higher level abstraction?
 
From: Alicia Drummond <adrum@microsoft.com> 
Sent: Wednesday, May 22, 2019 2:09 PM
To: Andrew Coates (REDMOND) <acoates@microsoft.com>; Harini Kannan <harinik@microsoft.com>
Cc: Keith Melmon <kmelmon@microsoft.com>; Stephen Peters <stpete@microsoft.com>
Subject: Re: onKey*Capture events
 
I got back from VAC today so I'm working through my backlog of emails.  I just figured I'd expand a little.

Event propagation is a concept common to many UI frameworks.  Capturing / At Target / Bubbling is the nomenclature used by Javascript in the Web.  In WPF XAML, these were called Tunnelling / Direct? / Bubbling.  As far as I know, the only UWP XAML example of a "Tunneling" event phase is PreviewKeyDown/Up like Andrew said which (I'm guessing) is implemented more like a Direct-only event but happens to occur in the Tunneling event phase of the Keyboard event.

Event propagation happens in both trees independently, and not all events are bubbled-- some are Direct-only.  XAML direct-only events don't have the RoutedEventArgs super class.  On the JS-side an event will be Direct-only if it's part of the GetExportedCustomDirectEventTypeConstants and Bubbled if GetExportedCustomBubblingEventTypeConstants with phasedRegistrationNames.  The pattern used by React Native is a little different from the web, here.  Bubbling events in Web JS just have an eventPhase field indicating the current phase and the same callback is used for all phases.  In React Native, the JS-side capturing phase goes through the onFooCapture callback (e.g. View has onMoveShouldSetResponder and onMoveShouldSetResponderCapture), but this is dependent on the native code setting up the phasedRegistrationNames properly... I think you could use a different name but it would break the pattern in the platform.  This is sort of how the Tunneling phase in UWP Xaml uses PreviewKeyDown instead of there being first class support for the Tunneling phase like in WPF.

The eventPhase enum I added was specifically for the declarative property to handle events, but I aligned with the names used in the Web.  So even though it's called Capturing, it's referring to the native-side's Capturing phase equivalent.  I.e. if you wanted to handle tab from the JS side instead of allowing the native platform to do so, you'd need to mark the event handled in PreviewKeyDown (the "tunneling/capturing" phase), but if you wanted to handle Space/Enter only if the native side hasn't (e.g. a dialog that dismisses on Enter, but with a Native button that handles Enter should take precedence if key focus is there) you'd want to mark the event handled in the bubbling phase (aka. The UIElement that is parent of the button's KeyDown).  My prop would look something like this:

<View keyDownEvents={{ key: 'Tab', eventPhase: EventPhase.Capturing }} onKeyDown={this._onKeyDown} />
<View keyDownEvents={{ key: 'Enter', eventPhase: EventPhase.Bubbling }} onKeyDown={this._onKeyDown} >
    <Button />
</View>
 
Alicia
________________________________________
